avl-tree/src/lib.rs

```diff
@@ -0,0 +1,589 @@
+                    Field::Left,
+                    self.get_field(leftmost, Field::Right),
+                );
+            }
+
+            self.set_field(leftmost, Field::Left, left);
+            if right != leftmost {
+                self.set_field(leftmost, Field::Right, right);
+            }
+
+            let (parent, branch, _) = path.pop().unwrap();
+
+            if let Some(parent) = parent {
+                self.set_field(parent, branch.unwrap(), leftmost);
+            }
+
+            path.push((parent, branch, leftmost));
+            if right != leftmost {
+                path.push((Some(leftmost), Some(Field::Right), right));
+            }
+            // drop the last inner_path element since it references the leftmost node
+            if !inner_path.is_empty() {
+                inner_path.pop();
+            }
+            path.extend(inner_path);
+
+            leftmost
+        } else {
+            let child = if left == SENTINEL && right == SENTINEL {
+                SENTINEL
+            } else {
+                if left != SENTINEL { left } else { right }
+            };
+
+            let (parent, branch, _) = path.pop().unwrap();
+
+            if let Some(parent) = parent {
+                self.set_field(parent, branch.unwrap(), child);
+
+                if child != SENTINEL {
+                    path.push((Some(parent), branch, child));
+                }
+            }
+
+            child
+        };
+
+        if node_index == self.root {
+            self.root = replacement;
+        }
+
+        self.delete(node_index);
+        self.rebalance(path);
+
+        Some(value)
+    }
+
+    fn balance_factor(&self, left: u32, right: u32) -> i32 {
+        // safe to convert to i32 since height will be at most log2(capacity)
+        let left_height = if left != SENTINEL {
+            self.get_field(left, Field::Height) as i32 + 1
+        } else {
+            0
+        };
+        let right_height = if right != SENTINEL {
+            self.get_field(right, Field::Height) as i32 + 1
+        } else {
+            0
+        };
+
+        left_height - right_height
+    }
+
+    fn left_rotate(&mut self, index: u32) -> u32 {
+        let right = self.get_field(index, Field::Right);
+        let right_left = self.get_field(right, Field::Left);
+
+        self.set_field(index, Field::Right, right_left);
+        self.set_field(right, Field::Left, index);
+
+        right
+    }
+
+    fn right_rotate(&mut self, index: u32) -> u32 {
+        let left = self.get_field(index, Field::Left);
+        let left_right = self.get_field(left, Field::Right);
+
+        self.set_field(index, Field::Left, left_right);
+        self.set_field(left, Field::Right, index);
+
+        left
+    }
+
+    fn update_height(&mut self, index: u32) {
+        let left = self.get_field(index, Field::Left);
+        let right = self.get_field(index, Field::Right);
+
+        let height = if (left + right) == 0 {
```

Comment:
Nit: I think this check makes more sense as `left == SENTINEL && right == SENTINEL` even though this is more verbose

Reply:
Done – now it is clearer the purpose of the comparison.